### PR TITLE
refactor: refactor to remove streams, emit info events

### PR DESCRIPTION
**Pull request by sveisvei**
## JIRA Issue
[COREWEB-107](https://jira.finn.no/browse/COREWEB-107)

## Status
**READY**

## Description
This PR should address code flow issues with streams complexity.

* emit info events 

BREAKING CHANGE: Drops support for sinks below v1.1
* upgraded @asset-pipe/sink-fs and @asset-pipe/sink-mem
* upgraded @asset-pipe/css-reader and @asset-pipe/js-reader

## Todos
- [x] Tests
- [x] Documentation

## Related PRs
* https://github.com/asset-pipe/asset-pipe-js-reader/pull/64
* https://github.com/asset-pipe/asset-pipe-css-reader/pull/9

### CODE REVIEW

File: lib/main.js [hunk 1]
```diff
@@ -172,83 +164,93 @@ module.exports = class Router extends EventEmitter {
...
-                    return next(
-                        boom.boomify(error, {
-                            statusCode: 400,
-                            message: 'Failed parsing postBundle POST-body',
-                        })
-                    );
+    postBundleHandler(type) {
+        assert(
+            ALLOWED_TYPES.includes(type),
+            `Expected type to 1 of ${ALLOWED_TYPES.join(', ')}, but got ${type}`
```
Comment (digitalsadhu) [inline lib/main.js]: s/to 1/to be 1

File: lib/main.js [hunk 2]
```diff
@@ -3,17 +3,27 @@
...
-const CssReader = require('asset-pipe-css-reader');
-const body = require('body/json');
+const bundleJS = require('@asset-pipe/js-reader');
+const bundleCSS = require('@asset-pipe/css-reader');
 const boom = require('boom');
 const uuid = require('uuid/v4');
 const mime = require('mime-types');
 const { Transform } = require('readable-stream');
+const { promisify } = require('util');
+const body = promisify(require('body/json'));
```
Comment (SimenB) [inline lib/main.js]: why? can't we just use `express.json()` middleware?

File: lib/main.js (hunk 1 above)
Comment (SimenB) [inline lib/main.js]: missing words in message

File: lib/main.js [hunk 3]
```diff
@@ -50,11 +60,14 @@ class CheckEmptyPayload extends Transform {
     }
 }
 
+const ALLOWED_TYPES = ['js', 'css'];
+
 module.exports = class Router extends EventEmitter {
-    constructor(sink) {
+    constructor(sink, logger = { info() {}, error() {} }) {
```
Comment (SimenB) [inline lib/main.js]: is the logger needed? we emit events now - why not stick to that pattern?

File: lib/main.js (hunk 2 above)
Comment (sveisvei) [reply lib/main.js]: It was there already...

File: lib/main.js [hunk 4]
```diff
@@ -172,83 +164,93 @@ module.exports = class Router extends EventEmitter {
...
+                            `Expected feed entry to be an array, but got ${typeof feed}`
+                        );
+                        return feed;
+                    })
+            )
+        );
 
-            reader.on('error', this.onError(next));
+        let bundle;
+        if (type === 'css') {
```
Comment (SimenB) [inline lib/main.js]: should we be explicit with `js` and throw if not one of those 2?

File: package.json [hunk 5]
```diff
@@ -76,10 +76,10 @@
     "clearMocks": true,
     "coverageThreshold": {
       "global": {
-        "branches": 100,
-        "functions": 100,
-        "lines": 100,
-        "statements": 100
+        "branches": 96,
```
Comment (SimenB) [inline package.json]: why? seems really unnecessary to lower coverage

File: lib/main.js (hunk 2 above)
Comment (SimenB) [reply lib/main.js]: express bundles body-parser now, no reason at all to not use it IMO. you already change its usage in this PR anyways

could have joi as middleware as well, but maybe not in this PR

**Comment by SimenB** [comment]
Should we add JOI schemas for the feeds?

File: lib/main.js [hunk 6]
```diff
@@ -172,83 +164,93 @@ module.exports = class Router extends EventEmitter {
...
+                    'request success',
+                    res.locals.track,
+                    req.method,
+                    req.path,
+                    response.file
+                );
+                res.json(response);
+            } catch (e) {
+                this.logger.error('Failed creating bundle', e);
+                if (e.message.includes('No file with name')) {
```
Comment (digitalsadhu) [inline lib/main.js]: I think we should probably have a better way to handle errors bubbling up from the sinks. Worth ignoring for now to get this out and revisiting afterwards though.

File: lib/main.js [hunk 7]
```diff
@@ -172,83 +164,93 @@ module.exports = class Router extends EventEmitter {
...
+        assert(
+            Array.isArray(feedIds) && feedIds.length > 0,
+            `Expected at least 1 feed id, but got ${feedIds}`
+        );
+        this.logger.info(`Processing feeds ${feedIds.join(', ')}`);
+        const feeds = await Promise.all(
+            feedIds.map(file =>
+                this.sink
+                    .get(file)
+                    .then(content => JSON.parse(content))
```
Comment (digitalsadhu) [inline lib/main.js]: .then(JSON.parse)

File: lib/main.js (hunk 4 above)
Comment (sveisvei) [reply lib/main.js]: it will throw higher up start-up-time.

File: package.json [hunk 8]
```diff
@@ -76,10 +76,10 @@
...
+        "functions": 99,
+        "lines": 99,
+        "statements": 99
```
Comment (digitalsadhu) [inline package.json]: WAT??? CHEATER!! ;)

File: package.json (hunk 5 above)
Comment (sveisvei) [reply package.json]: "Hard" error-state to reproduce in tests.

File: lib/main.js [hunk 9]
```diff
@@ -172,83 +164,93 @@ module.exports = class Router extends EventEmitter {
...
+                res.json(response);
+            } catch (e) {
+                this.logger.error('Failed creating bundle', e);
+                if (e.message.includes('No file with name')) {
                     return next(
-                        boom.boomify(error, {
-                            statusCode: 400,
-                            message: 'Invalid POST-body',
+                        boom.boomify(e, {
+                            statusCode: 404,
```
Comment (ivarconr) [inline lib/main.js]: Why is this a `404 Not-Found`?

File: lib/main.js [hunk 10]
```diff
@@ -172,83 +164,93 @@ module.exports = class Router extends EventEmitter {
...
-        return (req, res, next) => {
-            schemas.ids.validate(res.locals.payload, (error, payload) => {
-                if (error) {
+            try {
+                const response = await this.bundle({
+                    feedIds,
+                    uri: this.buildUri('bundle', req.headers.host, req.secure),
+                    type,
+                });
+                this.emit(
```
Comment (ivarconr) [inline lib/main.js]: why emit scuccess before `res.json` call?

Comment (SimenB) [reply lib/main.js]: used for logging (at least it was previously)

File: package.json (hunk 5 above)
Comment (SimenB) [reply package.json]: meh, not convinced :P

File: lib/main.js [hunk 11]
```diff
@@ -172,83 +164,93 @@ module.exports = class Router extends EventEmitter {
...
-            body(req, res, {}, (error, bodyObj) => {
-                if (error) {
-                    return next(
-                        boom.boomify(error, {
-                            statusCode: 400,
-                            message: 'Failed parsing postBundle POST-body',
-                        })
-                    );
+    postBundleHandler(type) {
+        assert(
```
Comment (ivarconr) [inline lib/main.js]: do we not catch the asset exception? Won't the server crash if the `assert` fails? Wouldn't it be better to return 400 bad request to the user? Or is there a global catch assert errors middleware handling this somewhere?

Comment (digitalsadhu) [reply lib/main.js]: This is a developer only error. Just a sanity check to make sure we used the method correctly. Once the server is running there would never be an error here.

File: lib/main.js (hunk 9 above)
Comment (digitalsadhu) [reply lib/main.js]: This error catches only when the sink file fetch cannot find a file. All other errors fall though. Do you think maybe thats wrong in the context of the whole route? ie. We request a bundle and get a 404 when any one of the feed files to be bundled cannot be found. I think that seems right?

File: lib/main.js (hunk 2 above)
Comment (digitalsadhu) [reply lib/main.js]: Yea I think we should revisit and use joi in a few places to validate feeds

File: lib/main.js (hunk 3 above)
Comment (digitalsadhu) [reply lib/main.js]: Now that you mention it, that does seem a bit inconsistent. Maybe sticking with emitting for now until we are clear about our logging/debugging story is wise.

File: lib/main.js (hunk 9 above)
Comment (ivarconr) [reply lib/main.js]: What about `409 Conflict`?

Comment (ivarconr) [reply lib/main.js]: Or maybe:

`412 Precondition Failed`
> The 412 (Precondition Failed) status code indicates that one or more
>  conditions given in the request header fields evaluated to false when
>  tested on the server.  This response code allows the client to place
>   preconditions on the current resource state (its current
>   representations and metadata) and, thus, prevent the request method
>   from being applied if the target resource is in an unexpected state.

File: lib/main.js (hunk 11 above)
Comment (ivarconr) [reply lib/main.js]: I was just  a bit trigger-happy and see now that these values are hard-coded and not part of the reques!

> Once the server is running there would __never__ be an error here.

:+1: Famous last words? :cocktail:

File: lib/main.js (hunk 2 above)
Comment (sveisvei) [reply lib/main.js]: was not drop-in-replacement as far as I can tell.

File: lib/main.js (hunk 11 above)
Comment (digitalsadhu) [reply lib/main.js]: Famous...last...words. 
So then goodbye cruel world...

File: lib/main.js (hunk 9 above)
Comment (sveisvei) [reply lib/main.js]: Change not related, but happy to change...

File: lib/main.js (hunk 7 above)
Comment (sveisvei) [reply lib/main.js]: fixing

File: lib/main.js (hunk 9 above)
Comment (digitalsadhu) [reply lib/main.js]: Does it have to be header fields? Looks like it maybe. In this case we are dealing with an array of hashes sent as a payload and 1 or more of them not being resolvable. Maybe the 409 option would be better?

File: lib/main.js (hunk 3 above)
Comment (sveisvei) [reply lib/main.js]: fixing...

File: lib/main.js (hunk 9 above)
Comment (sveisvei) [reply lib/main.js]: 409 it is then? @ivarconr

File: lib/main.js (hunk 2 above)
Comment (sveisvei) [reply lib/main.js]: https://github.com/asset-pipe/asset-pipe-build-server/issues/93

**Pull request merged**